### PR TITLE
Serialize dag timezone to a tzname or offset

### LINKED ISSUE #16613: InvalidTimezone exception if DAG's start_date timezone is "+00:00"
**Apache Airflow version**: 2.0.2

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`):

**Environment**:

- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:

```
from airflow.models import DAG
from airflow.serialization.serialized_objects import SerializedDAG
import pendulum
dag_start_date = pendulum.parse("2019-08-01T00:00:00.000+00:00")
dag = DAG(dag_id='simple_dag', start_date=dag_start_date)
serialized_dag = SerializedDAG.to_dict(dag)
serialized_dag['dag']['timezone'] # '+00:00'
dag = SerializedDAG.from_dict(serialized_dag) # raises InvalidTimezone exception

Traceback (most recent call last):
  File "/Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/pendulum/tz/zoneinfo/reader.py", line 50, in read_for
    file_path = pytzdata.tz_path(timezone)
  File "/Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/pytzdata/__init__.py", line 74, in tz_path
    raise TimezoneNotFound('Timezone {} not found at {}'.format(name, filepath))
pytzdata.exceptions.TimezoneNotFound: Timezone +00:00 not found at /Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/pytzdata/zoneinfo/+00:00
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3441, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-115-6c89c12cbb11>", line 1, in <module>
    dag = SerializedDAG.from_dict(serialized_dag)
  File "/Users/rubelagu/git/airflow/airflow/serialization/serialized_objects.py", line 795, in from_dict
    return cls.deserialize_dag(serialized_obj['dag'])
  File "/Users/rubelagu/git/airflow/airflow/serialization/serialized_objects.py", line 722, in deserialize_dag
    v = cls._deserialize_timezone(v)
  File "/Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/pendulum/tz/__init__.py", line 37, in timezone
    tz = _Timezone(name, extended=extended)
  File "/Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/pendulum/tz/timezone.py", line 40, in __init__
    tz = read(name, extend=extended)
  File "/Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/pendulum/tz/zoneinfo/__init__.py", line 9, in read
    return Reader(extend=extend).read_for(name)
  File "/Users/rubelagu/.pyenv/versions/airflow-venv/lib/python3.8/site-packages/pendulum/tz/zoneinfo/reader.py", line 52, in read_for
    raise InvalidTimezone(timezone)
pendulum.tz.zoneinfo.exceptions.InvalidTimezone: Invalid timezone "+00:00"

```

**What you expected to happen**:

The DAG holds a reference to the DAG's `start_date.tzinfo` and it will serialize  as `+00:00` (this can be checked with `serialized_dag['dag']['timezone']`, then when it's time to deserialize that it will try to do `pendulum.timezone('+00:00`)` which raises a `InvalidTimezone` exception. 

In principle I would expect to be able to provide any `datetime` as start_date , and `+00:00` is common. The serialization/deserialization will be used in normal airflow operation so that any DAG with that kind of start_date will give exceptions. 

Probably the **`dag.timezone` should not be serialized at all** and it should be **reconstructed** at deserialization time from `start_date`.

Refactor this [section of airflow/models/dag.py::DAG.__init__()](https://github.com/apache/airflow/blob/c2af5e3ca22eca7d4797b141520a97cf5e5cc879/airflow/models/dag.py#L304-L313) into a method that can be called from both `DAG.__init__` and `SerializedDAG.from_dict`. That way the problem of serialize/deserialize a pendulum.timezone would be avoided.

**How to reproduce it**:

**Anything else we need to know**:

Related issue: #16551
Related PR: #16599

### PR DESCRIPTION
**Pull request by ecerulm**
closes: #16613
related: #16631

Tries to serialize the `dag.timezone` in a slightly more robust way, using
* use `start_date.tzinfo.name`  if exist and it's recognized by pendulum as valid timezone identifier
* use `start_date.tzinfo.tzname(...)` if it's recognized by pendulum as valid timezone identifier
* use the `start_date.tzinfo.utcoffset(...)`  as last resort.

This allows to use `DAG(....,start_date=pendulum.parse("2021-06-01T12:00:00+01:00"))` which was not possible before because the timezone was serialized as `+01:00` and it raised an `InvalidTimezone` exception at deserialization time when tried to execute `pendulum.timezone('+01:00')`

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by ecerulm** [comment]
There is another PR (also my me) that tries to solve the problem #16631 more at the deserialization time.

File: airflow/models/dag.py [hunk 1]
```diff
@@ -422,6 +428,38 @@ def __exit__(self, _type, _value, _tb):
 
     # /Context Manager ----------------------------------------------
 
+    def _get_tzname_or_offset(self, obj):
```
Comment (uranusjr) [inline airflow/models/dag.py]: This function (and `_is_valid_pendulum_tz_id`) does not use anything on the DAG (i.e. `self`) and should be free functions instead. Maybe put them at the top of the module or in `airflow.utils.dates`?

File: airflow/models/dag.py [hunk 2]
```diff
@@ -422,6 +428,38 @@ def __exit__(self, _type, _value, _tb):
...
+        try:
+            candidate = int(obj.utcoffset(datetime.now(timezone.utc)).total_seconds())
+            if self._is_valid_pendulum_tz_id(candidate):
+                return candidate
+        except (NotImplementedError, ValueError):
+            pass
+
+        raise ValueError(f"Can't get a timezone name or offset from {obj}")
+
+    def _is_valid_pendulum_tz_id(self, id):
```
Comment (uranusjr) [inline airflow/models/dag.py]: Dom’t name a variable `id` (linters will be unhappy). “Tz id” is also a bit confusing to me (that value does not actually identify anything—id is short for identifier). Maybe call it `name` or something?

File: airflow/models/dag.py [hunk 3]
```diff
@@ -302,16 +304,20 @@ def __init__(
...
             if isinstance(self.default_args['start_date'], str):
                 self.default_args['start_date'] = timezone.parse(self.default_args['start_date'])
-            self.timezone = self.default_args['start_date'].tzinfo
+            if self.default_args['start_date'] and self.default_args['start_date'].tzinfo:
+                self.schedule_timezone = self._get_tzname_or_offset(self.default_args['start_date'].tzinfo)
 
-        if not hasattr(self, 'timezone') or not self.timezone:
-            self.timezone = settings.TIMEZONE
+        if not hasattr(self, 'schedule_timezone'):
+            self.schedule_timezone = self._get_tzname_or_offset(settings.TIMEZONE)
```
Comment (uranusjr) [inline airflow/models/dag.py]: This `hasattr` call is super weird. This is the initialiser, we should have a way to know whether `self.schedule_timezone` is set or not without introspection. (Granted the existing code already does the same, but that does not make it less awkward.)

**Review by uranusjr — COMMENTED** [review]
It’s not clear to me why a new argument is needed to solve this. From what I can tell, this PR has two distinct parts, 1. make the DAG’s timezone configuration explicit (instead of relying on `start_date`), and 2. correctly solve timezone coersion so it can always be correctly serialised to a deserialised value. The latter is the bug we originally intend to solve, and while the former is also a good idea, it is not required to fix the original problem we have, and the problem’s solution is the same regardless of it. So while personally I think it’s a good idea to make the timezone configuration explicit, it should probably be discussed separately (maybe a short AIP would be needed), and not as a part of the fix.

Some thoughts on the implementation below.

**Comment by ecerulm** [comment]
>It’s not clear to me why a new argument is needed to solve this.
>....
> and the problem’s solution is the same regardless of it.

well that's exactly the issue, The **partial** solution is the same, but the I believe that the true only solution is to make the user specify the timezone as as a string or int  instead of trying to get it as `datetime.tzinfo` (of pendulum's subclasses).  But, yes, you are right, there is no need to introduce that new argument just to have an equivalent of the  #16631 implementation

I just tried to create an AIP at https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals  and I don't have permissions to do it.

Anyway, I'll just wait to see @ashb  comments as well, because first somebody needs to decide if we want to handle the problem in the DAG.__init__() , so that serialize/deserialize is simpler like in this PR, or handle it at deserialization time like in #16631.

File: airflow/models/dag.py [hunk 4]
```diff
@@ -302,16 +304,20 @@ def __init__(
         self.fileloc = back.f_code.co_filename if back else ""
         self.task_dict: Dict[str, BaseOperator] = {}
 
-        # set timezone from start_date
+        # set timezone from schedule_timezone or start_date
+        if schedule_timezone:
+            self.schedule_timezone = self._get_tzname_or_offset(schedule_timezone)
```
Comment (ashb) [inline airflow/models/dag.py]: Should we do any checks to make sure that the start_date is in the right timezone?

(Just trying to think about potential foot-guns that might surprise users.)

File: tests/serialization/test_dag_serialization.py [hunk 5]
```diff
@@ -135,7 +136,7 @@
                 'label': 'custom_task',
             },
         ],
-        "timezone": "UTC",
+        "schedule_timezone": "UTC",
```
Comment (ashb) [inline tests/serialization/test_dag_serialization.py]: Okay, so now we come to an interesting problem that we, so far, have not had to deal with.

We need to be able to continue to parse "old" serialized versions, so this might be easiest to _not_ change the field name, and in the serialized representation keep it called `timezone`, but just upgrade the type to allow it to become int or string.

File: airflow/models/dag.py (hunk 4 above)
Comment (uranusjr) [reply airflow/models/dag.py]: +1 on checking `start_date` is either naive (and will be converted to `schedule_timezone`) or matches `schedule_timezone`. Maybe `end_date` as well, although we don’t currently do? (Which is still a footgun.)

Comment (ecerulm) [reply airflow/models/dag.py]: I don't think we really want to check that the `start_date` time zone matches the `schedule_timezone`.  That's why I wanted to introduce `schedule_timezone` to avoid having to rely on inferring things from `start_date.tzinfo`.

There is no problem to specify the `start_date` in one timezone, the `end_date` in another timezone, and have yet another timezone for the `schedule_timezone`.  I can see people (myself included) using  iso 8601 string like `2021-06-01T12:00:00+01:00` for `start_date` and `schedule_timezone='Europe/Stockholm'`

Comment (ecerulm) [reply airflow/models/dag.py]: But anyway, I'm removing `schedule_timezone` from the PR, right? Because that changes the public api of the DAG and that requires an AIP, right?

Comment (uranusjr) [reply airflow/models/dag.py]: > There is no problem to specify the `start_date` in one timezone, the `end_date` in another timezone, and have yet another timezone for the `schedule_timezone`. I can see people (myself included) using iso 8601 string like `2021-06-01T12:00:00+01:00` for `start_date` and `schedule_timezone='Europe/Stockholm'`

Hmm, that makes sense. And since if `schedule_timezone` is _not_ set, `start_date.tzinfo` will be used anyway, there is no behavioural change for people not (yet) modifying their DAGs. And when `schedule_timezone` is set, the user should know what they’re doing.

> But anyway, I'm removing `schedule_timezone` from the PR, right? Because that changes the public api of the DAG and that requires an AIP, right?

I think so, yes, but we’re going to need to have this discussion when explicit per-DAG timezone setup becomes a thing (whether via `schedule_timezone` or another mean) anyway.

File: airflow/models/dag.py (hunk 1 above)
Comment (ecerulm) [reply airflow/models/dag.py]: Ok I moved to the module level

File: airflow/models/dag.py (hunk 2 above)
Comment (ecerulm) [reply airflow/models/dag.py]: ok renamed it to `is_valid_pendulum_tzname` and `name`

File: airflow/models/dag.py (hunk 4 above)
Comment (ecerulm) [reply airflow/models/dag.py]: Ok, removed `schedule_timezone` from DAG constructor constructor

File: tests/serialization/test_dag_serialization.py (hunk 5 above)
Comment (ecerulm) [reply tests/serialization/test_dag_serialization.py]: Ok, I managed to keep as `timezone` in the serialized object, but this getting more and more complex , and more hacky that the alternative #16631.

**Comment by ecerulm** [comment]
So I removed the `schedule_timezone` argument and rename  the `schedule_timezone ` attribute to `_schedule_timezone`(private) as both things were chaing the public api of the DAG and an AIP it's needed to modify the AIP.

@potiuk, @ashb  what is required to write an AIP? I don't have the permissions to do it.

File: tests/serialization/test_dag_serialization.py (hunk 5 above)
Comment (ashb) [reply tests/serialization/test_dag_serialization.py]: If we can find a to not have to do `return pendulum.parse("2021-01-01T12:00:00" + tz_name).tzinfo` I'd be happy with that approach, but with that it just feels too hacky I think.

File: airflow/models/dag.py [hunk 6]
```diff
@@ -107,6 +108,40 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
     return query.first()
 
 
+def _get_tzname_or_offset(obj):
```
Comment (ashb) [inline airflow/models/dag.py]: Should live in airflow.utils.timezone

File: airflow/models/dag.py [hunk 7]
```diff
@@ -107,6 +108,40 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+    except (NotImplementedError, ValueError):
+        pass
+
+    raise ValueError(f"Can't get a timezone name or offset from {obj}")
+
+
+def _is_valid_pendulum_tzname(name):
+    if isinstance(name, str) or isinstance(name, int):
+        try:
+            return pendulum.timezone(name)
```
Comment (ashb) [inline airflow/models/dag.py]:
```suggestion
            pendulum.timezone(name)
            return True
```

File: airflow/models/dag.py [hunk 8]
```diff
@@ -107,6 +108,40 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+        if _is_valid_pendulum_tzname(candidate):
+            return candidate
+    except (NotImplementedError, ValueError):
+        pass
+
+    raise ValueError(f"Can't get a timezone name or offset from {obj}")
+
+
+def _is_valid_pendulum_tzname(name):
+    if isinstance(name, str) or isinstance(name, int):
```
Comment (ashb) [inline airflow/models/dag.py]:
```suggestion
    if isinstance(name, (str, int)):
```

File: airflow/models/dag.py [hunk 9]
```diff
@@ -304,14 +339,15 @@ def __init__(
...
             if isinstance(self.default_args['start_date'], str):
                 self.default_args['start_date'] = timezone.parse(self.default_args['start_date'])
-            self.timezone = self.default_args['start_date'].tzinfo
+            if self.default_args['start_date'] and self.default_args['start_date'].tzinfo:
+                self._schedule_timezone = _get_tzname_or_offset(self.default_args['start_date'].tzinfo)
 
-        if not hasattr(self, 'timezone') or not self.timezone:
-            self.timezone = settings.TIMEZONE
+        if not hasattr(self, '_schedule_timezone'):
+            self._schedule_timezone = _get_tzname_or_offset(settings.TIMEZONE)
```
Comment (ashb) [inline airflow/models/dag.py]: Does any of this actualy need changing?

Can't we keep `self.timezone` as an actual timezone object?

File: airflow/models/dag.py [hunk 10]
```diff
@@ -2209,7 +2249,7 @@ def get_num_task_instances(dag_id, task_ids=None, states=None, session=None):
     def get_serialized_fields(cls):
         """Stringified DAGs and operators contain exactly these fields."""
         if not cls.__serialized_fields:
-            cls.__serialized_fields = frozenset(vars(DAG(dag_id='test')).keys()) - {
+            cls.__serialized_fields = frozenset(set(vars(DAG(dag_id='test')).keys())) - {
```
Comment (ashb) [inline airflow/models/dag.py]: Unnecessary change if we keep time timezone as an actual property.

File: airflow/serialization/serialized_objects.py [hunk 11]
```diff
@@ -186,7 +186,9 @@ def serialize_to_json(
             if cls._is_excluded(value, key, object_to_serialize):
                 continue
 
-            if key in decorated_fields:
+            if key == 'timezone':
+                serialized_object[key] = cls._serialize(object_to_serialize._schedule_timezone)
+            elif key in decorated_fields:
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]: We shouldn't need this here, handle it in `def _serialize` instead.

File: airflow/serialization/serialized_objects.py [hunk 12]
```diff
@@ -712,7 +714,8 @@ def deserialize_dag(cls, encoded_dag: Dict[str, Any]) -> 'SerializedDAG':
                 v = {task["task_id"]: SerializedBaseOperator.deserialize_operator(task) for task in v}
                 k = "task_dict"
             elif k == "timezone":
-                v = cls._deserialize_timezone(v)
+                dag._schedule_timezone = v
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]: Not needed if we keep `dag.timezone` as an attribute.

**Review by ashb — CHANGES_REQUESTED** [review]
I think you can simplufy this PR a lot, and isolate the change to _just_ serialization.

The dag object itself doesn't ever need the timezone _name_, it just needs the tz object.

So if we use the logic you have got for getting the TZ name or UTC offset when we serialize a TZ object, then the existing deserialize logic would work as is.

**Comment by ecerulm** [comment]
I moved the functions to the `airflow.utils.timezone` removed all changes from `airflow.models.dag` and keep everything in `airflow.serializaton.serialized_objects`

File: airflow/models/dag.py (hunk 9 above)
Comment (ecerulm) [reply airflow/models/dag.py]: sure

File: airflow/models/dag.py (hunk 3 above)
Comment (ecerulm) [reply airflow/models/dag.py]: I left dag.py untouched now

File: airflow/utils/timezone.py [hunk 13]
```diff
@@ -184,3 +185,38 @@ def coerce_datetime(v: Union[None, dt.datetime, DateTime]) -> Optional[DateTime]
...
+            return candidate
+    try:
+        candidate = obj.tzname(dt.datetime.now(dt.timezone.utc))
+        if _is_valid_pendulum_tzname(candidate):
+            return candidate
+    except (NotImplementedError, ValueError):
+        pass
+
+    try:
+        candidate = int(obj.utcoffset(dt.datetime.now(dt.timezone.utc)).total_seconds())
```
Comment (ecerulm) [inline airflow/utils/timezone.py:205]: @ashb , @uranusjr do we want to do this or not? at this point there is no real guarantee that the `datetime.tzinfo` that is really a fixed offset timezone. There is no good way to check if it really is. I can do an `isinstance(obj, pendulum.tz.timezone.FixedTimezone)` that works only for pendulum timezones.

 The risk is that someone uses some custom `datetime.tzinfo` with a custom name with no equivalent in pendulum, and this serialization logic will reduce that custom tzinfo to a fixedoffset timezone (which will have a different behaviour than the original dag.timezone).

Comment (ashb) [reply airflow/utils/timezone.py:205]: > The risk is that someone uses some custom datetime.tzinfo with a custom name with no equivalent in pendulum

I think so long as we document "don't do this"  I'm okay wit this. The chance of someone creating a custom timezone seems _very_ slim given Airflow's typical users.

Comment (ashb) [reply airflow/utils/timezone.py:205]: https://airflow.apache.org/docs/apache-airflow/stable/timezone.html?highlight=timezone#time-zone-aware-dags would be the place to put such a comment.

File: airflow/serialization/serialized_objects.py [hunk 14]
```diff
@@ -27,6 +27,8 @@
 import pendulum
 from dateutil import relativedelta
 
+from airflow.utils.timezone import _get_tzname_or_offset
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]:
```suggestion
from airflow.utils.timezone import get_tzname_or_offset
```

If it's called from outside the module it shouldn't have `_` prefix.

File: airflow/serialization/serialized_objects.py [hunk 15]
```diff
@@ -228,7 +230,7 @@ def _serialize(cls, var: Any) -> Any:  # Unfortunately there is no support for r
         elif isinstance(var, datetime.timedelta):
             return cls._encode(var.total_seconds(), type_=DAT.TIMEDELTA)
         elif isinstance(var, Timezone):
-            return cls._encode(str(var.name), type_=DAT.TIMEZONE)
+            return cls._encode(_get_tzname_or_offset(var), type_=DAT.TIMEZONE)
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py]:
```suggestion
            return cls._encode(get_tzname_or_offset(var), type_=DAT.TIMEZONE)
```

**Review by ashb — COMMENTED** [review]
Looks good to me now.

Any comments @uranusjr ?

File: airflow/utils/timezone.py (hunk 13 above)
Comment (ecerulm) [reply airflow/utils/timezone.py:205]: Ok, I added a "Don't use other `datetime.tzinfo` implementations" caveat on that page (timezone.rst)

**Comment by uranusjr** [comment]
Nothing more from me for now, but one thing for the future:

> > The risk is that someone uses some custom datetime.tzinfo with a custom name with no equivalent in pendulum
> 
> I think so long as we document "don't do this" I'm okay wit this. The chance of someone creating a custom timezone seems _very_ slim given Airflow's typical users.

We’ll probably need to revisit this when 3.10 is out (and supported by Airflow) since that will become the most obvious tzinfo implementation instead of `pendulum.Timzeon`. A follow-up issue tracking `zoneinfo.Zoneinfo` compatibility (and potentially an eventual switch to it from pendulum) would be nice to have.

**Comment by ashb** [comment]
Oh, do we actually need it to be a pendulum.TZ, or do we already "support" that and the build in dt.zoneinfo?

**Comment by ecerulm** [comment]
> Oh, do we actually need it to be a pendulum.TZ, or do we already "support" that and the build in dt.zoneinfo?

There is no guarantee that he zoneinfo tzname() will return something that can be fed to `pendulum.timezone(xx)` 

```
from zoneinfo import ZoneInfo
from datetime import datetime,timedelta
import pendulum
dt = datetime(2020, 10, 31, 12, tzinfo=ZoneInfo("America/Los_Angeles"))
print(dt.tzname()) # PDT
pendulum.timezone(dt.tzname()) # InvalidTimezone "PDT"
```

So for example `ZoneInfo("America/Los_Angeles")` will be interpreted as an integer offset in the current status of this PR. 

If we wanted to support `ZoneInfo` too we would need to check for `.key` attribute too (in addition to `.name`) and hopefully they both have the same IANA names supported so `pendulum.timezone(zinfoobj.key)` will provide an equivalent pendulum timezone to the specified zoneinfo. 

But it's not better to just say that it has to be a pendulum timezone, since at the end at deserializaiton time it will be reconstructed as a pendulum?

**Comment by uranusjr** [comment]
Yeah, `tzinfo.tzname()` is not suitable for our serialisation use since it makes zero guarantee about the result’s decodability or round-trip property. The possible fix to this would be pretty wide in scope (we’ll probably need an tzinfo implementation interoperability standard) and let’s not go there for now…

File: airflow/serialization/serialized_objects.py [hunk 16]
```diff
@@ -284,7 +286,7 @@ def _deserialize(cls, encoded_var: Any) -> Any:
         elif type_ == DAT.TIMEDELTA:
             return datetime.timedelta(seconds=var)
         elif type_ == DAT.TIMEZONE:
-            return Timezone(var)
+            return pendulum.timezone(var)
```
Comment (ashb) [inline airflow/serialization/serialized_objects.py:289]: For consistency with _deserilalized_dag, should this be

```suggestion
            return self._deserialize_timezone(var)
```

(I think the functionality is identical.)

Comment (kaxil) [reply airflow/serialization/serialized_objects.py:289]: @ecerulm ^^ can you take a look at this suggestion please and rebase to the latest main at the same time, thanks
